### ISSUE TITLE
Frame synchronization in Metal

### ISSUE THREAD

**Issue by kvark** [gfx-rs/gfx#2143]
After https://github.com/gfx-rs/gfx/pull/2138 got in, we got a problem with exclusive full-screen mode of Dota, where a semaphore is provided for the frame, and there is nothing instantly available. So `nextDrawable == nil`, and we don't get a `CADrawable`, and we don't get the associated `MTLTexture`... which is needed for immediate command buffer recording.

One possible solution would be to switch to deferred recording once we encounter a texture that will be available later (from an acquired frame). We'd only be able to record the rest when we get that texture.

**Comment by kvark** [comment]
This is a real blocker/PITA. MoltenVK isn't affected that much because they defer the actual command recording to the submission time, where they are safe to wait for semaphores first, and thus they have all the required textures to work with.

As for my proposed workaround, we don't know out of hand where exactly such a texture would be needed. Say, we are in the middle of a render pass, and receive a `bind_graphics_descriptor_sets` call that has a reference to a frame. If we switch to deferred recording, we'd have to keep a hold on that WIP render pass in the mean time. So it still seems possible, but not super convenient or easy.

**Comment by msiglreith** [comment]
For clarification:
`acquire_image -> record commands -> record command with image with current acquired image -> submit cmd buffer (including wait on acquire)`
The issue here is that the image is not ready but it's needs to be for recording already?

**Comment by kvark** [comment]
@msiglreith correct, yes. When recording a command buffer immediately (i.e. one-time submit ones), we need `MTLTexture`, but we can only get it from `CADrawable`, which we can only get *after* the frame-sync wait.
